File: findFirstLast.py
"""
Finding first and last occurence of target given.
Time = O(log n)
space = O(1)
Successfully excecuted on leetcode

Approach:
1. Two binary search implementation for finding first and last occurence
Left Boundry ---
2. If mid < target, then the range must begins on the right of mid, so l = mid+1
3. If mid > target, it means the range must begins on the left of mid (r = mid-1)
3. If mid = target, then the range must begins on the left of or at mid (r= mid)
Right Boundry --
4. If mid > target, then the range must begins on the left of mid (r = mid-1)
5. If mid < target, then the range must begins on the right of mid, so l = mid+1 
6. If mid = target, then the range must begins on the right of or at mid (l= mid)


"""

class FindFirstLast:
    
    def findFL(self, nums, target):
        if not nums:
            return [-1,-1]
        n = len(nums)
        left, right = -1, -1
        l, r = 0, n-1
        while l < r:
            mid = (l+r)//2
            if nums[mid] < target:
                l = mid+1
            else: 
                r = mid
        if nums[l] != target: 
            return -1, -1
        left = l
        l, r = left, n-1
        while l < r:
            mid = (l+r)//2+1 #plus 1 to make mid biased towards right
            if nums[mid] == target: 
                l = mid
            else: 
                r = mid-1
        right = l
        return left, right

File: Find_first_last.java
// Time Complexity : O(log(n))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public int[] searchRange(int[] nums, int target) {
        if(nums==null||nums.length==0)
            return new int[]{-1,-1};

        int [] ans = new int[]{-1,-1};
        ans[0] = first(nums,target);
        ans[1] = last(nums,target);
        return ans;
    }

    int first(int[] nums, int target){
        int l =0;
        int h = nums.length-1;
        while(l<=h){
            int mid = l +(h-l)/2;
            if(nums[mid]<target)
                l = mid+1;
            else if(nums[mid]>target)
                h = mid-1;
            else {
                if(mid==0 || nums[mid-1]!=target){
                    return mid;
                }
                else{
                    h= mid-1;
                }
            }
        }
        return -1;
    }
    int last(int[] nums, int target){
        int l = 0;
        int h = nums.length-1;
        while(l<=h){
            int mid = l +(h-l)/2;
            if(nums[mid]<target)
                l = mid+1;
            else if(nums[mid]>target)
                h = mid-1;
            else{
                if(mid==nums.length-1 || nums[mid+1]!=target){
                    return mid;
                }
                else{
                    l = mid+1;
                }
            }
        }
        return -1;
    }
}